### PR TITLE
Adding support for multi-node DDP training

### PR DESCRIPTION
**Pull request by smedhe**
Add support for multi-node Distributed Data Parallel (DDP) training to the QEfficient finetuning pipeline. This enables scaling training across multiple nodes while keeping the existing single-node behavior unchanged.

Commands for DDP across 2 servers:
For the Master Addr or the Primary Machine, use node-rank as 0:
QAIC_VISIBLE_DEVICES=0,1,2,3 torchrun --nnodes=2 --nproc-per-node=4 --seed 0 --node-rank=0 --master_addr=<MASTER_NODE_IP> --master_port=8000 -m QEfficient.cloud.finetune --device qaic --enable_ddp --model_name "meta-llama/Llama-3.2-1B" --dataset alpaca_dataset --train_batch_size 1 --val_batch_size 1 --num_epochs 1 --max_train_step 200 --max_eval_step 50

For Node 1, use node-rank as 1:
QAIC_VISIBLE_DEVICES=0,1,2,3 torchrun --nnodes=2 --nproc-per-node=4 --seed 0 --node-rank=1 --master_addr=<MASTER_NODE_IP> --master_port=8000 -m QEfficient.cloud.finetune --device qaic --enable_ddp --model_name "meta-llama/Llama-3.2-1B" --dataset alpaca_dataset --train_batch_size 1 --val_batch_size 1 --num_epochs 1 --max_train_step 200 --max_eval_step 50

### CODE REVIEW

File: QEfficient/cloud/finetune.py [hunk 1]
```diff
@@ -67,32 +66,60 @@ def setup_distributed_training(train_config: TrainConfig) -> None:
...
+
     dist_backend_map = {"cpu": "gloo", "qaic": "qccl", "cuda": "gloo"}
-    dist.init_process_group(backend=dist_backend_map[torch_device.type])
+    dist.init_process_group(dist_backend_map[torch_device.type], rank=rank, world_size=world_size)
+
+    # Set the base device index for this process on this node
+    # For PP: each process controls num_pp_stages devices starting from base_device_index
+    base_device_index = local_rank * train_config.num_pp_stages
     # from here onward "qaic/cuda" will automatically map to "qaic:i/cuda:i", where i = process rank
-    getattr(torch, torch_device.type).set_device(dist.get_rank() * train_config.num_pp_stages)
```
Comment (quic-akuruvil) [inline QEfficient/cloud/finetune.py]: Test whether this change would affect the PP pipeline.

File: QEfficient/finetune/utils/helper.py [hunk 2]
```diff
@@ -47,13 +47,12 @@ def enum_names(enum_cls: Enum) -> List[str]:
 
 
 def get_rank() -> int:
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/helper.py]: Please add another method ( get_local_rank() )which returns local rank which was earlier returned by the current method.

File: QEfficient/cloud/finetune.py [hunk 3]
```diff
@@ -67,32 +66,60 @@ def setup_distributed_training(train_config: TrainConfig) -> None:
...
         return
 
+    # ---- DDP path (single- or multi-node) ----
     assert torch_device.type != "cpu", "Host doesn't support single-node DDP"
-    assert torch_device.index is None, f"DDP requires only device type, got: {torch_device}"
+    assert torch_device.index is None, f"DDP requires only device type (qaic/cuda), got: {torch_device}"
+
+    # Torchrun-provided env vars
+    world_size = int(os.environ.get("WORLD_SIZE", "1"))
+    rank = int(os.environ.get("RANK", "0"))
```
Comment (quic-swatia) [inline QEfficient/cloud/finetune.py]: Why are we not fetching it from the method present in helper file? Please explain if I am missing something here. 
IMO rank and world_size should be fetched directly from the helper methods while there should be methods defined for local_rank and local_world_size inside the helper.py

File: QEfficient/cloud/finetune.py [hunk 4]
```diff
@@ -67,32 +66,60 @@ def setup_distributed_training(train_config: TrainConfig) -> None:
...
+    local_rank = int(os.environ.get("LOCAL_RANK", "0"))
+    local_world_size = int(os.environ.get("LOCAL_WORLD_SIZE", "1"))
+
+    # Sanity check vs helper
+    assert world_size == get_world_size(), f"WORLD_SIZE={world_size} but get_world_size()={get_world_size()}"
```
Comment (quic-swatia) [inline QEfficient/cloud/finetune.py]: both world_size and get_world_size() are assigned os.environ.get("WORLD_SIZE", "1"), so it's not required.

File: QEfficient/finetune/utils/helper.py (hunk 2 above)
Comment (smedhe) [reply QEfficient/finetune/utils/helper.py]: Resolved.

**Comment by quic-swatia** [comment]
> For PP, device_map.py uses get_rank() which was earlier returning local rank but now returns rank. This will not break PP as of now, but it's a wrong practice. In case PP is used with multi node, it would break. Please call the get_local_rank() in device_map once it is defined.

After this change, we can also test if PP + DDP for multi node is working with this PR. It seems like it might work. It's okay if it doesn't, we can still go ahead with the PR.

**Review by quic-meetkuma — APPROVED** [review]
Looks fine to me. Please update the finetune.md file. Imagine or other teams refer the commands from there. They will need to change command at their side of the code.

File: QEfficient/cloud/finetune.py (hunk 4 above)
Comment (quic-akuruvil) [reply QEfficient/cloud/finetune.py]: @smedhe Please push the changes and then mark as resolve

**Review by quic-akuruvil — COMMENTED** [review]
Start CI run also for this. So CI passes before merge
